### ISSUE TITLE
Add base types to proposal

### ISSUE THREAD

**Issue by Naddiseo** [lukescott/es-type-hinting#32]
Mention the adding of "number", "string", "boolean" as (contextual?) reserved words as a difference from the current spec. Define the semantics behind them (if any?).

**Comment by lukescott** [comment]
Reserved in the sense that they will be used in the future sure. Use them outside of TypeHint, probably not until they've been reserved for a while.

**Comment by Naddiseo** [comment]
I'm thinking they should be reserved in the same way "async" is; they can still be used as variable/function names. Hm, but that will lead to the rare possibility of:

``` javascript
class number {}
function fn() number {} // returns <class number> or <type number> ?
```

**Comment by lukescott** [comment]
That's where this gets a bit tricky. I would think that `number` would be special in a TypeHint and ignore that number class. It would probably have to be since `number` would be `undefined`. That could be confusing though.

We could define `number`, similar to how [types-objects](https://github.com/dslomov-chromium/typed-objects-es7) will define new types. But then you will have to reserve those or allow them to be shadowed. The typed-objects proposal, in its current form, would have to solve this problem too.

You had brought up using strings in one of the other issues. We could possibly use strings for the base types, such as:

``` javascript
function fn() "number" {}
class Foo {}
function fn2() Foo {}
```

But that feels a little dirty to me. And if a string were allowed you could put just about anything in it.

I think `number` should be a special word only usable in type hints though:

``` javascript
class number {} // should be invalid
class class {} // just like this is invalid
```

Although there are some compatibility issues with that approach.

**Comment by Naddiseo** [comment]
From a simplicity perspective, it's easier to go with option 1: it's interpreted as the TypeHint not as the class in that context. That way the spec is cleaner, and the shadowing can be left up to a linter.

**Comment by lukescott** [comment]
Would the same also be true for `Foo`? What about runtime reflection? Does that return a reference or a string (`"number` or `"Foo"`)?

**Comment by Naddiseo** [comment]
Yeah, the reflection is going to make this harder. This is what I'm thinking:

``` javascript
function %SetHint(ref, hint) {
switch (%StringValue(hint)) {
    case "string": // base type
       ref[@@hint] = Reflect.type.string; break;
    case "number": // base type
        ref[@@hint] = Reflect.type.number; break;
    case "boolean": // base type
         ref[@@hint] = Reflect.type.boolean; break;
    case "void": // base type
         ref[@@hint] = Reflect.type.undefined; break;
    default: // User defined
        ref[@@hint] = Reflect.type.object; // maybe use `hint` instead of `Reflect.type.object`
        ref[@@userhint] = hint; // or something, 
}
}

function fn() string { }
// desugars:
function fn() {}
%SetHint(fn, @@stringType);

function fn() UserValue {}
// desugars:
function fn() {}
%SetHint(fn, UserValue);

```

Return type can be reflected:
`retType = fn[@@hint];`

**Comment by lukescott** [comment]
That kind of assumes `string`, `number`, etc. are actual reference types, doesn't it? Unless you did something like:

``` javascript
function fn() string {}
// desugars:
function fn() {}
%SetHint(fn, "string");
```

**Comment by Naddiseo** [comment]
Yeah. You have to assume they're reference types of some kind (or a symbol maybe) to be able to accommodate reflection.

**Comment by lukescott** [comment]
Ok, so assuming they are references and you have this:

``` javascript
function fn() string {}
// desugars:
function fn() {}
%SetHint(fn, string);
```

What does this produce:

``` javascript
console.log(string);
```

Is `string` a class of some sort?

**Comment by Naddiseo** [comment]
I think for the base types we'd have to introduce some global primitive objects, like what the reflection proposal adds with `@@symbolType` (or use theirs).

**Comment by lukescott** [comment]
Does typed objects in #31 have a symbol type that we can use? It has Type Objects. Would that qualify?
